valuable/src/pointer.rs

```diff
@@ -0,0 +1,205 @@
+//! References to fields of [`Valuable`] values.
+//!
+//! A [`Pointer`] stores a path traversal to a particular value in a nested
+//! [`Valuable`] structure. For example, a [`Pointer`] might refer to the field `z`
+//! of the field `y` of the value `x`. [`Pointer`] paths can also include indices into
+//! [`Listable`] and [`Tuplable`] values, in order to represent expressions like
+//! `x.y[3].0`.
+//!
+//! # Examples
+//!
+//! ```rust
+//! use valuable::*;
+//!
+//! #[derive(Valuable)]
+//! struct Struct1 {
+//!     x: String,
+//!     y: Struct2,
+//! }
+//!
+//! #[derive(Valuable)]
+//! struct Struct2 {
+//!     z: String,
+//! }
+//!
+//! struct Visitor;
+//!
+//! impl Visit for Visitor {
+//!     fn visit_value(&mut self, value: Value<'_>) {
+//!         println!("{:?}", value);
+//!     }
+//! }
+//!
+//! let value = Struct1 {
+//!     x: "a".to_owned(),
+//!     y: Struct2 {
+//!         z: "b".to_owned(),
+//!     },
+//! };
+//!
+//! let mut visitor = Visitor;
+//!
+//! visit_pointer!(value.x, visitor);   // "a"
+//! visit_pointer!(value.y, visitor);   // Struct2 { field: "b" }
+//! visit_pointer!(value.y.z, visitor); // "b"
+//! ```
+//!
+//! [`Listable`]: crate::Listable
+//! [`Tuplable`]: crate::Tuplable
+
+use crate::{NamedValues, Slice, Valuable, Value, Visit};
+
+/// A pointer to the value.
+#[derive(Debug, Clone, Copy)]
```

Comment:
it might be nice to have a `Display` formatter that formats the `Pointer` as an expression, like the inputs to the `visit_pointer!` macro. but, we could add that in a follow-up PR.